File: client/lua/krpc/encoder.lua
local pb = require 'protobuf.pb'
local tablex = require 'pl.tablex'
local schema = require 'krpc.schema.KRPC'
local Types = require 'krpc.types'

local encoder = {}

-- Where the routines below leave the bytes they produce. They are handed to a writer, and one
-- writer shared by all of them costs nothing per value, where a closure made for the value
-- would be allocated and thrown away again for every one encoded.
local _written
local function _write(data)
  _written = data
end

-- What pb.struct_pack calls a single precision and a double precision number.
local FLOAT_FORMAT = string.byte('f')
local DOUBLE_FORMAT = string.byte('d')

-- The encoding of every varint that fits in one byte, which is most of the values a call
-- carries: an index, a count, a small identifier or a boolean.
local _small_varints = {}
for value = 0, 127 do
  _small_varints[value] = string.char(value)
end

local function _encode_varint(x)
  local small = _small_varints[x]
  if small then
    return small
  end
  if x < 0 then
    error('Value must be non-negative, got ' .. x)
  elseif x == math.huge then
    return '\255\255\255\255\255\255\255\255\127'
  else
    pb.varint_encoder(_write, x)
    return _written
  end
end

local function _encode_float(value)
  pb.struct_pack(_write, FLOAT_FORMAT, value)
  return _written
end

local function _encode_double(value)
  pb.struct_pack(_write, DOUBLE_FORMAT, value)
  return _written
end

-- How to encode a value, by the code of the type it has. Which of these a type wants is the
-- first thing encoding a value asks, and a table says so in one lookup where asking a type what
-- class it is walks its ancestry once per question.
local _value_encoders = {
  [Types.DOUBLE] = _encode_double,
  [Types.FLOAT] = _encode_float,
  [Types.SINT32] = function(x) return _encode_varint(pb.zig_zag_encode32(x)) end,
  [Types.SINT64] = function(x) return _encode_varint(pb.zig_zag_encode64(x)) end,
  [Types.UINT32] = _encode_varint,
  [Types.UINT64] = _encode_varint,
  [Types.BOOL] = function(x) return _encode_varint(x and 1 or 0) end,
  [Types.STRING] = function(x) return _encode_varint(x:len()) .. x end,
  [Types.BYTES] = function(x) return _encode_varint(x:len()) .. x end,
  [Types.ENUMERATION] = function(x) return _encode_varint(pb.zig_zag_encode32(x.value)) end,
  [Types.CLASS] = function(x) return _encode_varint(x._object_id) end,
}

function encoder.encode(x, typ)
  local code = typ.code
  local encode_value = _value_encoders[code]
  if encode_value then
    return encode_value(x)
  end
  if code == Types.LIST then
    local msg = schema.List()
    for item in x:iter() do
      msg.items:append(encoder.encode(item, typ.value_type))
    end
    return msg:SerializeToString()
  elseif code == Types.DICTIONARY then
    local msg = schema.Dictionary()
    for key,value in tablex.sort(x) do
      local entry = msg.entries:add()
      entry.key = encoder.encode(key, typ.key_type)
      entry.value = encoder.encode(value, typ.value_type)
    end
    return msg:SerializeToString()
  elseif code == Types.SET then
    local msg = schema.Set()
    for item in pairs(x) do
      msg.items:append(encoder.encode(item, typ.value_type))
    end
    return msg:SerializeToString()
  elseif code == Types.TUPLE then
    local msg = schema.Tuple()
    for _,item in ipairs(tablex.zip(x, typ.value_types)) do
      msg.items:append(encoder.encode(item[1], item[2]))
    end
    return msg:SerializeToString()
  elseif typ:is_a(Types.MessageType) then
    return x:SerializeToString()
  end
  error('Cannot encode object of type ' .. tostring(typ))
end

function encoder.encode_message_with_size(message)
  -- Encode a message prefixed by its size
  local data = message:SerializeToString()
  local delimiter = _encode_varint(data:len())
  return delimiter .. data
end

return encoder

File: client/python/krpc/encoder.py
from __future__ import annotations
from typing import cast, Any, Callable, Collection, Iterable, List, Mapping
from enum import Enum
import struct

# pylint: disable=import-error,no-name-in-module
import google.protobuf
from google.protobuf.internal import encoder as protobuf_encoder

# pylint: disable=import-error,no-name-in-module
from google.protobuf.internal import wire_format as protobuf_wire_format
from krpc.error import EncodingError
from krpc.platform import bytelength
import krpc.schema.KRPC_pb2 as KRPC
from krpc.types import (
    Types,
    TypeBase,
    ValueType,
    ClassType,
    EnumerationType,
    MessageType,
    StructType,
    TupleType,
    ListType,
    SetType,
    DictionaryType,
)

# The following unpacks the internal protobuf decoders, whose signature
# depends on the version of protobuf installed
_pb_VarintEncoder = protobuf_encoder._VarintEncoder()  # type: ignore[attr-defined]  # pylint: disable=invalid-name
_pb_SignedVarintEncoder = protobuf_encoder._SignedVarintEncoder()  # type: ignore[attr-defined]  # pylint: disable=invalid-name

# protobuf's zigzag encoder carries no types, so it is bound here with the type it does
# have. A call carrying a collection reaches it once per value, so this also saves looking
# it up on its module that many times
_pb_zigzag_encode: Callable[[int], int] = protobuf_wire_format.ZigZagEncode

# The one byte encoding of every value that fits in one. A varint below 128 is the byte
# itself, and message sizes and the length prefixes of strings and bytes almost always
# are, so the common case is a lookup
_SINGLE_BYTE_VARINTS = tuple(bytes((value,)) for value in range(128))


class Encoder:
    """Routines for encoding messages and values in
    the protocol buffer serialization format"""

    _types = Types()

    @classmethod
    def encode(cls, x: object, typ: TypeBase) -> bytes:
        """Encode a message or value of the given protocol buffer type.

        This encodes a value at a slot, where a null is carried by the is_null flag of the
        message around it. A value at a position inside another value carries its own
        presence bool, which _item_encoder writes."""
        # The value types come first, as most arguments are one, and this is on the hot
        # path of every remote procedure call
        if isinstance(typ, ValueType):
            return cls._encode_value(x, typ)
        if isinstance(typ, MessageType):
            return cast(google.protobuf.message.Message, x).SerializeToString()
        if isinstance(typ, EnumerationType):
            return cls._encode_value(cast(Enum, x).value, cls._types.sint32_type)
        if isinstance(typ, ClassType):
            object_id = x._object_id  # type: ignore[attr-defined]
            return cls._encode_value(object_id, cls._types.uint64_type)
        if isinstance(typ, ListType):
            list_msg = KRPC.List()
            encode_item = cls._item_encoder(typ.value_type)
            list_msg.items.extend(
                encode_item(item) for item in cast(Iterable[object], x)
            )
            return list_msg.SerializeToString()
        if isinstance(typ, DictionaryType):
            dict_msg = KRPC.Dictionary()
            entries = []
            dict_obj = cast(Mapping, x)  # type: ignore[type-arg]
            encode_key = cls._item_encoder(typ.key_type)
            encode_value = cls._item_encoder(typ.value_type)
            for key, value in sorted(
                dict_obj.items(), key=lambda i: i[0]
            ):  # type: ignore[no-any-return]
                entry = KRPC.DictionaryEntry()
                entry.key = encode_key(key)
                entry.value = encode_value(value)
                entries.append(entry)
            dict_msg.entries.extend(entries)
            return dict_msg.SerializeToString()
        if isinstance(typ, SetType):
            set_msg = KRPC.Set()
            encode_item = cls._item_encoder(typ.value_type)
            set_msg.items.extend(
                encode_item(item) for item in cast(Iterable[object], x)
            )
            return set_msg.SerializeToString()
        if isinstance(typ, TupleType):
            tuple_msg = KRPC.Tuple()
            tuple_obj = cast(Collection[object], x)
            if len(tuple_obj) != len(typ.value_types):
                raise EncodingError(
                    "Tuple has wrong number of elements. "
                    + "Expected %d, got %d." % (len(typ.value_types), len(tuple_obj))
                )
            tuple_msg.items.extend(cls._encode_items(tuple_obj, typ.value_types))
            return tuple_msg.SerializeToString()
        if isinstance(typ, StructType):
            # A structure is encoded as the values of its fields in order, which is the
            # same encoding as a tuple of those values
            struct_msg = KRPC.Tuple()
            struct_obj = cast(Collection[object], x)
            if len(struct_obj) != len(typ.field_types):
                raise EncodingError(
                    "Struct has wrong number of fields. "
                    + "Expected %d, got %d." % (len(typ.field_types), len(struct_obj))
                )
            struct_msg.items.extend(cls._encode_items(struct_obj, typ.field_types))
            return struct_msg.SerializeToString()
        raise EncodingError("Cannot encode objects of type " + str(type(x)))

    @classmethod
    def encode_message_with_size(
        cls, message: google.protobuf.message.Message
    ) -> bytes:
        """Encode a protobuf message, prepended with its size"""
        data = message.SerializeToString()
        length = len(data)
        if length < 128:
            return _SINGLE_BYTE_VARINTS[length] + data
        size: bytes = protobuf_encoder._VarintBytes(length)  # type: ignore[attr-defined]
        return size + data

    @classmethod
    def _encode_items(
        cls, values: Iterable[object], types: Iterable[TypeBase]
    ) -> List[bytes]:
        """The encoded value of each item of a tuple or a structure.

        A tuple is encoded on the hot path of every remote procedure call. Extending the
        protobuf repeated field from a list costs less than from a generator."""
        return [cls._item_encoder(typ)(value) for value, typ in zip(values, types)]

    @classmethod
    def _item_encoder(cls, typ: TypeBase) -> Callable[[Any], bytes]:
        """A function that encodes one value at a position of the given type"""
        encode = cls._value_encoder(typ)
        if not typ.nullable:
            return encode
        # A nullable position carries a presence bool before its value, and holds nothing
        # else when that bool is false
        present = cls._encode_value(True, cls._types.bool_type)
        absent = cls._encode_value(False, cls._types.bool_type)
        return lambda x: absent if x is None else present + encode(x)

    @classmethod
    def _value_encoder(cls, typ: TypeBase) -> Callable[[Any], bytes]:
        """A function that encodes one value of the given type.

        A collection carries many values of the same type, so working out how to encode
        one of them is worth doing once for the collection rather than once for every
        item in it. The types a collection usually holds are answered directly; anything
        else falls back to the full encode."""
        if isinstance(typ, ValueType):
            encode = _VALUE_ENCODERS.get(typ.code)
            if encode is None:
                raise EncodingError("Invalid type")
            return encode
        if isinstance(typ, ClassType):
            encode_id = _VALUE_ENCODERS[cls._types.uint64_type.code]
            return lambda x: encode_id(x._object_id)
        if isinstance(typ, EnumerationType):
            encode_value = _VALUE_ENCODERS[cls._types.sint32_type.code]
            return lambda x: encode_value(x.value)
        return lambda x: cls.encode(x, typ)

    @classmethod
    def _encode_value(cls, value: object, typ: TypeBase) -> bytes:
        encode = _VALUE_ENCODERS.get(typ.code)
        if encode is None:
            raise EncodingError("Invalid type")
        return encode(value)


class _ValueEncoder:
    """Routines for encoding values in the
    protocol buffer serialization format"""

    @classmethod
    def encode_double(cls, value: float) -> bytes:
        return struct.pack("<d", value)

    @classmethod
    def encode_float(cls, value: float) -> bytes:
        return struct.pack("<f", value)

    @classmethod
    def _encode_varint(cls, value: int) -> bytes:
        if value < 128:
            return _SINGLE_BYTE_VARINTS[value]
        data: List[bytes] = []
        _pb_VarintEncoder(data.append, value, True)
        return b"".join(data)

    @classmethod
    def _encode_signed_varint(cls, value: int) -> bytes:
        value = _pb_zigzag_encode(value)
        if value < 128:
            return _SINGLE_BYTE_VARINTS[value]
        # Two bytes carry a value up to 16383, which covers most values above one byte, and
        # writing them costs one call where protobuf's encoder builds a list and joins it.
        # A collection reaches this once per value it carries.
        if value < 16384:
            return bytes((value & 0x7F | 0x80, value >> 7))
        data: List[bytes] = []
        _pb_SignedVarintEncoder(data.append, value, True)
        return b"".join(data)

    @classmethod
    def encode_uint32(cls, value: int) -> bytes:
        if value < 0:
            raise EncodingError("Value must be non-negative, got %d" % value)
        return cls._encode_varint(value)

    @classmethod
    def encode_uint64(cls, value: int) -> bytes:
        if value < 0:
            raise EncodingError("Value must be non-negative, got %d" % value)
        return cls._encode_varint(value)

    @classmethod
    def encode_bool(cls, value: bool) -> bytes:
        return cls._encode_varint(value)

    @classmethod
    def encode_string(cls, value: str) -> bytes:
        size = cls._encode_varint(bytelength(value))
        data = value.encode("utf-8")
        return size + data

    @classmethod
    def encode_bytes(cls, value: bytes) -> bytes:
        return b"".join([cls._encode_varint(len(value)), value])


# The encoder for each value type, looked up by type code, as this is on the hot path of
# every remote procedure call. The signed integer types share an encoder, named for what
# it writes
_VALUE_ENCODERS: Mapping[KRPC.Type.TypeCode, Callable[[Any], bytes]] = {
    KRPC.Type.SINT32: _ValueEncoder._encode_signed_varint,
    KRPC.Type.SINT64: _ValueEncoder._encode_signed_varint,
    KRPC.Type.UINT32: _ValueEncoder.encode_uint32,
    KRPC.Type.UINT64: _ValueEncoder.encode_uint64,
    KRPC.Type.DOUBLE: _ValueEncoder.encode_double,
    KRPC.Type.FLOAT: _ValueEncoder.encode_float,
    KRPC.Type.BOOL: _ValueEncoder.encode_bool,
    KRPC.Type.STRING: _ValueEncoder.encode_string,
    KRPC.Type.BYTES: _ValueEncoder.encode_bytes,
}
